c_src/otter_nif.cpp

```diff
@@ -199,35 +277,82 @@ static bool get_args_with_type(ErlNifEnv *env, ERL_NIF_TERM arg_types_term, std:
+    if (wrapper_it != struct_type_wrapper_registry.end()) {
+        return &wrapper_it->second;
+    }
+    FFIStructTypeWrapper wrapper_;
+    auto insert_it = struct_type_wrapper_registry.insert({struct_id, std::move(wrapper_)});
+    if (!insert_it.second) return nullptr;
+    auto& wrapper = insert_it.first->second;
+    wrapper.struct_id = struct_id;
+    wrapper.resource_type = get_ffi_struct_resource_type(env, struct_id);
+    if (get_args_with_type(env, array[2], args_with_type)) {
+        for (auto& p : args_with_type ) {
+            if (p.second == "c_ptr") {
+                wrapper.field_types.push_back(&ffi_type_pointer);
+            } else {
+                wrapper.field_types.push_back(str2ffi_type[p.second]);
+            }
+        }
+        wrapper.finalize();
+        return &wrapper;
+    }
+    return nullptr;
+}
+
 static ERL_NIF_TERM otter_invoke(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]) {
     if (argc != 3) return enif_make_badarg(env);
 
     OtterSymbol * symbol_res;
+
     std::string return_type;
+    FFIStructTypeWrapper* struct_return_type;
+
     std::vector<std::pair<ERL_NIF_TERM, std::string>> args_with_type;
-    static std::map<std::string, ffi_type *> str2ffi_type = {
-        {"u8", &ffi_type_uint8},
-        {"u16", &ffi_type_uint16},
-        {"u32", &ffi_type_uint32},
-        {"u64", &ffi_type_uint64},
-        {"s8", &ffi_type_sint8},
-        {"s16", &ffi_type_sint16},
-        {"s32", &ffi_type_sint32},
-        {"s64", &ffi_type_sint64},
-        {"f32", &ffi_type_float},
-        {"f64", &ffi_type_double},
-    };
-
-    if (enif_get_resource(env, argv[0], OtterSymbol::type, (void **)&symbol_res) &&
-        erlang::nif::get_atom(env, argv[1], return_type) &&
-        get_args_with_type(env, argv[2], args_with_type)) {
+    if (erlang::nif::get_atom(env, argv[1], return_type)) {
+        // Do nothing
+    }
+    else if (auto created = FFIStructTypeWrapper::create_from_tuple(env, argv[1])) {
+        struct_return_type = created;
+    }
+    else {
+        return erlang::nif::error(env, "fail to get return_type");
+    }
+    if (enif_get_resource(env, argv[0], OtterSymbol::type, (void **)&symbol_res)) {
+        if (!get_args_with_type(env, argv[2], args_with_type)) return erlang::nif::error(env, "fail to get args_with_type");
         void * symbol = symbol_res->val;
         if (symbol != nullptr) {
             ffi_cif cif;
-            ffi_type **args = (ffi_type **)enif_alloc(sizeof(ffi_type *) * args_with_type.size());
-            void ** values = (void **)enif_alloc(sizeof(void *) * args_with_type.size());
-            ffi_type * ffi_return_type;
-            void * rc = malloc(sizeof(ffi_arg));
+            // NOTE: These values are small structs, allocating them on stack should be ok
+            ffi_type *args[args_with_type.size()];
+            void * values[args_with_type.size()];
+            ffi_type * ffi_return_type = nullptr;
+            ffi_arg rc;
```

Comment:
We need to allocate memory for `rc`, otherwise, the call stack could be overwritten by the return value if the size of the return value is larger than `sizeof(ffi_arg)`.

Reply:
I know. But before changing rc I encountered something strange and it is keep causing segfault. I will write it down here.